src/main/kotlin/org/jitsi/nlj/rtp/ResumableStreamRewriter.kt

```diff
@@ -0,0 +1,166 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.nlj.rtp
+
+import org.jitsi.rtp.extensions.bytearray.putInt
+import org.jitsi.rtp.rtcp.RtcpSrPacket
+import org.jitsi.rtp.rtp.RtpPacket
+import org.jitsi.rtp.util.isNewerThan
+import org.jitsi.rtp.util.isNewerTimestampThan
+
+/**
+ * Rewrites sequence numbers for RTP streams by hiding any gaps caused by
+ * dropped packets. Rewriters are not thread-safe. If multiple threads access a
+ * rewriter concurrently, it must be synchronized externally.
+ *
+ * Port of the class in libjitsi.
+ *
+ * @author Maryam Daneshi
+ * @author George Politis
+ * @author Boris Grozev
+ */
+class ResumableStreamRewriter {
+    /**
+     * The sequence number delta between what's been accepted and what's been
+     * received, mod 2^16.
+     */
+    var seqnumDelta = 0
+        private set
+
+    /**
+     * The timestamp delta between what's been accepted and what's been
+     * received, mod 2^32.
+     */
+    private var timestampDelta: Long = 0
+
+    /**
+     * The highest sequence number that got accepted, mod 2^16.
+     */
+    var highestSequenceNumberSent = -1
+        private set
+
+    /**
+     * The highest timestamp that got accepted, mod 2^32.
+     */
+    private var highestTimestampSent: Long = -1
+
+    /**
+     * Rewrites the sequence number of the given RTP packet hiding any gaps caused by drops.
+     *
+     * @param accept true if the packet is accepted, false otherwise
+     */
+    fun rewriteRtp(accept: Boolean, rtpPacket: RtpPacket) {
+        val sequenceNumber = rtpPacket.sequenceNumber
+        val newSequenceNumber = rewriteSequenceNumber(accept, sequenceNumber)
+
+        val timestamp = rtpPacket.timestamp
+        val newTimestamp = rewriteTimestamp(accept, timestamp)
+
+        if (sequenceNumber != newSequenceNumber) {
+            rtpPacket.sequenceNumber = newSequenceNumber
+        }
+
+        if (timestamp != newTimestamp) {
+            rtpPacket.timestamp = newTimestamp
+        }
+    }
+
+    /**
+     * Restores the RTP timestamp of the RTCP SR packet in the buffer.
+     *
+     */
+    fun rewriteRtcpSr(rtcpPacket: RtcpSrPacket) {
+        if (timestampDelta == 0L) {
+            return
+        }
+
+        val timestamp = rtcpPacket.senderInfo.rtpTimestamp
+        val newTimestamp = (timestamp - timestampDelta) and 0xffffffffL
+
+        // TODO: what's the right way to do this?
+        //rtcpPacket.senderInfo.rtpTimestamp = newTimestamp
```

Comment:
is this something we've got to get figured out before merging?

Reply:
The current solution is a total hack from the point of view of architecture. But it works well and quick and it doesn't add much debt (it's a one liner), so I was hoping we could get it merged this way and improve later (I'm happy to add a task to the board). I don't really know what the proper solution would look like at this point.

Reply:
ok, yeah sound like we should add a task to make sure we revisit